crates/derive/src/stages/frame_queue.rs

```diff
@@ -47,9 +49,94 @@ where
     /// Create a new [FrameQueue] stage with the given previous [L1Retrieval] stage.
     ///
     /// [L1Retrieval]: crate::stages::L1Retrieval
-    pub fn new(prev: P) -> Self {
+    pub fn new(prev: P, cfg: Arc<RollupConfig>) -> Self {
         crate::set!(STAGE_RESETS, 0, &["frame-queue"]);
-        Self { prev, queue: VecDeque::new() }
+        Self { prev, queue: VecDeque::new(), rollup_config: cfg }
+    }
+
+    /// Returns if holocene is active.
+    pub fn is_holocene_active(&self) -> bool {
+        let Some(origin) = self.prev.origin() else {
+            warn!(target: "frame-queue", "Failed to get origin");
+            return false;
+        };
+        self.rollup_config.is_holocene_active(origin.timestamp)
+    }
+
+    /// Prunes frames if Holocene is active.
+    pub fn prune(&mut self) {
+        if !self.is_holocene_active() {
+            return;
+        }
+
+        let mut i = 0;
+        while i < self.queue.len() - 1 {
```

Comment:
Hmm.. It seems like looping through the full queue every time (we call `self.prune()` every `load_frames` call post-holocene is a bit inefficient. Can we move towards only pruning when the new frame comes in?

It looks like [the spec](https://specs.optimism.io/protocol/holocene/derivation.html#frame-queue) mentions that most of these cases where we drop a frame happen atomically as a frame is absorbed, i.e. the "non-first frame" cases.

For the "first frame, where prev.is_last == false" case, we should be able to just drain the full queue and add the new first frame, without having to iterate through every single one of them.

Reply:
Following up on this